## Random Forest test

In [7]:
import pydotplus 
from IPython.display import Image
import numpy as np
import pandas
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestClassifier
import os
from collections import defaultdict
import _pickle as cPickle
from scipy.stats import kurtosis, skew
%matplotlib notebook

In [2]:
def preprocessing(data):
    result = []
    mean = np.mean(data)
    median = np.median(data)
    maximum = np.amax(data)
    minimum = np.amin(data)
    std_dev = np.std(data)
    var = np.var(data)
    ran = np.ptp(data)
    skewness = skew(data)
    kurto = kurtosis(data)
    result=[mean, median, maximum, minimum,
                   std_dev, var, ran, skewness, kurto]
    return result

In [3]:
dirname = './data_preprocessed_python'
filenames = os.listdir(dirname)
data_dic = defaultdict(lambda:[])
for i, filename in enumerate(filenames):
    full_filename = os.path.join(dirname, filename)
    x = cPickle.load(open(full_filename, 'rb'), encoding='ISO-8859-1')
    data_dic[filename[:-4]] = x

In [5]:
new_data_dic = defaultdict(lambda:defaultdict(lambda:[]))
for key in data_dic.keys():
    par = int(key[1:])
    
    for i, vid in enumerate(data_dic[key]['data']):
        new_data_dic[key]['data'].append([])
        for chnn in vid[:-8]:
            l =preprocessing(np.array(chnn))
            for batch in range(0,10):
                l = l+preprocessing(np.array(chnn[batch*807:max((batch+1)*807, 8064)]))
            new_data_dic[key]['data'][i].append(np.array([par, i+1]+l))
    for i,  vid in enumerate(data_dic[key]['labels']):
        l2 = []
        new_data_dic[key]['labels'].append([])
        for value in vid:
            if value>=5:
                l2.append(1)
            else:
                l2.append(0)
        new_data_dic[key]['labels'][i] += l2

In [6]:
train_data = []
train_label = []
test_data = []
test_label = []
for i in new_data_dic:
    print("process {}".format(i))
    if i != 's02':
        
        for j in range(len(new_data_dic[i]['labels'])):
            temp = np.array(new_data_dic[i]['data'][j]).reshape(-1, 32*101)[0]
            train_data.append(temp)
            train_label.append(new_data_dic[i]['labels'][j])
    else:
        for j in range(len(new_data_dic[i]['labels'])):
            temp = np.array(new_data_dic[i]['data'][j]).reshape(-1, 32*101)[0]
            test_data.append(temp)
            test_label.append(new_data_dic[i]['labels'][j])
print("Finished!")

process s23
process s18
process s13
process s27
process s31
process s16
process s28
process s12
process s04
process s19
process s25
process s10
process s17
process s06
process s30
process s09
process s05
process s20
process s07
process s01
process s15
process s22
process s02
process s03
process s11
process s29
process s21
process s24
process s26
process s32
process s14
process s08
Finished!


In [8]:
dt = RandomForestClassifier(n_estimators=100, oob_score=True, random_state=123456, criterion='entropy')

# fit it to our data
dt.fit(train_data, train_label)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=1,
            oob_score=True, random_state=123456, verbose=0,
            warm_start=False)

### The accuracy of Random Forest was the highest.

In [11]:
result = []
for i in range(len(test_data)):
    result.append(dt.predict(test_data[i].reshape(1,-1))[0])

count = 0
for index, i in enumerate(result):
    for j in range(len(i)):
        if(i[j]==test_label[index][j]):
            count += 1
print("The Accuracy is: {}%".format(count/(len(test_label)*4) * 100))

The Accuracy is: 72.5%


In [12]:
new_data_dic_3 = defaultdict(lambda:defaultdict(lambda:[]))
for key in data_dic.keys():
    par = int(key[1:])
    
    for i, vid in enumerate(data_dic[key]['data']):
        new_data_dic_3[key]['data'].append([])
        for chnn in vid[:-8]:
            l =preprocessing(np.array(chnn))
            for batch in range(0,10):
                l = l+preprocessing(np.array(chnn[batch*807:max((batch+1)*807, 8064)]))
            new_data_dic_3[key]['data'][i].append(np.array([par, i+1]+l))
    for i,  vid in enumerate(data_dic[key]['labels']):
        l2 = []
        new_data_dic_3[key]['labels'].append([])
        for value in vid:
            if value>=6:
                l2.append(2)
            if value<6 and value >= 4:
                l2.append(1)
            if value < 4:
                l2.append(0)
        new_data_dic_3[key]['labels'][i] += l2

In [14]:
train_data_3 = []
train_label_3 = []
test_data_3 = []
test_label_3 = []
for i in new_data_dic:
    print("process {}".format(i))
    if i != 's02':
        for j in range(len(new_data_dic_3[i]['labels'])):
            temp = np.array(new_data_dic_3[i]['data'][j]).reshape(-1, 32*101)[0]
            train_data_3.append(temp)
            train_label_3.append(new_data_dic_3[i]['labels'][j])
    else:
        for j in range(len(new_data_dic_3[i]['labels'])):
            temp = np.array(new_data_dic_3[i]['data'][j]).reshape(-1, 32*101)[0]
            test_data_3.append(temp)
            test_label_3.append(new_data_dic_3[i]['labels'][j])
print("Finished!")

process s23
process s18
process s13
process s27
process s31
process s16
process s28
process s12
process s04
process s19
process s25
process s10
process s17
process s06
process s30
process s09
process s05
process s20
process s07
process s01
process s15
process s22
process s02
process s03
process s11
process s29
process s21
process s24
process s26
process s32
process s14
process s08
Finished!


In [16]:
dt_3 = RandomForestClassifier(n_estimators=100, oob_score=True, random_state=123456, criterion='entropy')

# fit it to our data
dt_3.fit(train_data_3, train_label_3)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=1,
            oob_score=True, random_state=123456, verbose=0,
            warm_start=False)

In [17]:
result = []
for i in range(len(test_data_3)):
#     print(dt_3.predict(test_data_3[i].reshape(1,-1))[0])
    result.append(dt_3.predict(test_data_3[i].reshape(1,-1))[0])

count = 0
for index, i in enumerate(result):
    for j in range(len(i)):
        if(i[j]==test_label_3[index][j]):
            count += 1
print("The Accuracy is: {}%".format(count/(len(test_label_3)*4) * 100))

The Accuracy is: 50.0%


### Result
- 2 Classes accuracy 72.5%
- 3 Classes accuracy 50%

- Best algorithm in this experiment!